src/modules/rlm_ldap/rlm_ldap.c

```diff
@@ -949,18 +1015,64 @@ static int rlm_ldap_groupcmp(void *instance, request_t *request, fr_pair_t const
 	return 1;
 }
 
+/** Perform async lookup of user DN if required for authentication
+ *
+ */
+static unlang_action_t mod_authenticate_start(UNUSED rlm_rcode_t *p_result, UNUSED int *priority,
+					      request_t *request, void *uctx)
+{
+	ldap_auth_ctx_t	*auth_ctx = talloc_get_type_abort(uctx, ldap_auth_ctx_t);
+
+	return rlm_ldap_find_user_async(auth_ctx, auth_ctx->inst, request, &auth_ctx->mod_env->user_base,
+					&auth_ctx->mod_env->user_filter, auth_ctx->ttrunk, NULL, NULL);
+}
+
+/** Initiate async LDAP bind to authenticate user
+ *
+ */
+static unlang_action_t mod_authenticate_resume(rlm_rcode_t *p_result, UNUSED int *priority,
+					       request_t *request, void *uctx)
+{
+	ldap_auth_ctx_t	*auth_ctx = talloc_get_type_abort(uctx, ldap_auth_ctx_t);
+
+	/*
+	 *	Arriving here from an LDAP search will mean the dn in auth_ctx is NULL.
+	 */
+	if (!auth_ctx->dn) auth_ctx->dn = rlm_find_user_dn_cached(request);
+
+	/*
+	 *	No DN found - can't authenticate the user.
+	 */
+	if (!auth_ctx->dn) {
+	fail:
+		talloc_free(auth_ctx);
+		RETURN_MODULE_FAIL;
+	}
+
+	/*
+	 *	Attempt a bind using the thread specific connection for bind auths
+	 */
+	if (auth_ctx->mod_env->user_sasl_mech.type == FR_TYPE_STRING) {
+		ldap_auth_mod_env_t *mod_env = auth_ctx->mod_env;
+		if (fr_ldap_sasl_bind_auth_async(request, auth_ctx->thread, mod_env->user_sasl_mech.vb_strvalue,
+						 auth_ctx->dn, mod_env->user_sasl_authname.vb_strvalue,
+						 auth_ctx->password, mod_env->user_sasl_proxy.vb_strvalue,
+						 mod_env->user_sasl_realm.vb_strvalue) < 0) goto fail;
+	} else {
+		if (fr_ldap_bind_auth_async(request, auth_ctx->thread, auth_ctx->dn, auth_ctx->password) < 0) goto fail;
+	}
+
+	RETURN_MODULE_RCODE(unlang_interpret_synchronous(unlang_interpret_event_list(request), request));
```

Comment:
Guessing this is temporary? :)

Reply:
Yes, until the bind auth is switched to use a trunk so it can be properly async